azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py

```diff
@@ -273,45 +276,77 @@ def _on_provider_message_received_callback(self, topic, payload):
             pass  # is there any other case
 
     def _on_provider_unsubscribe_complete(self, mid):
-        """
-        Callback that is called by the provider when a subscribe operation is complete.
-        """
-        logger.info("_on_provider_unsubscribe_complete")
-        callback = self._unsubscribe_callback
-        if callback:
-            self._unsubscribe_callback = None
+        if mid in self._callback_map:
+            callback = self._callback_map[mid]
+            del self._callback_map[mid]
             callback()
+        else:
+            # TODO: tests for unkonwn MID cases
+            logger.warning("UNSUBACK received with unknown MID: %s", str(mid))
 
-    def _add_event_to_queue(self, event_data):
+    def _add_action_to_queue(self, event_data):
         """
-        Queue an event for sending later.  All events that get sent end up going into
-        this queue, even if they're going to be sent immediately.
+        Queue an action for running later.  All actions that need to run while connected end up in
+        this queue, even if they're going to be run immediately.
         """
-        logger.info("Adding event to queue for later sending")
-        # TODO: throw here if event_data.args[0] is not an event/message
-        self._event_queue.put_nowait((event_data.args[0], event_data.args[1]))
+        action = event_data.args[0]
+        if isinstance(action, TransportAction):
+            self._action_queue.put_nowait(event_data.args[0])
+        else:
+            assert False
```

Comment:
>assert False [](start = 12, length = 12)

I'm not sure I understand how we could end up here. if the goal is to provoke a crash that will encourage a customer to publish a bug report, i think we should add a debug statement for that "Oops! looks like you found a bug - please submit a github issue <...instructions...>"

Reply:
Adding onto this, why don't we raise an Exception here instead? After all, "assert False" raises an AssertionError, which isn't terribly informative (and is intended largely for testing) - we might as well choose what gets raised.

---
In reply to: [266184228](https://github.com/Azure/azure-iot-sdk-python-preview/pull/63#discussion_r266184228) [](ancestors = 266184228)